File: cs445/threading/src/gus/worker/Producer.java
package gus.worker;

public class Producer extends Thread {
  private String name;
  private Consumer consumer;
  private int consumablesToCreate = 0;

  public Producer(String name, Consumer consumer, int consumablesToCreate) {
    this.name = name;
    this.consumer = consumer;
    this.consumablesToCreate = consumablesToCreate;
    start();
  }

  public void run() {
    System.out.println("started " + name);
    try {
      for (int i = 0; i < consumablesToCreate; i++) {
        sleep(i % 2);
        consumer.addConsumable(name + "-" + String.valueOf(i));
      }
    } catch (InterruptedException e) {}
    System.out.println("stopped " + name);
  }

  public static void main (String [] args) {
    Consumer consumer = new Consumer();
    Thread consumerThread = new Thread(consumer);
    consumerThread.start();
    
    new Producer("happy", consumer, 5);
    new Producer("dopey", consumer, 8);
    new Producer("skidmark", consumer, 10);
  }
}

File: cs445/threading/ruby/worker/producer.rb
require 'worker/consumer'

#
# ruby worker/producer.rb
class Producer
  def initialize(name, consumer, consumables_to_create)
    @name, @consumer, @consumables_to_create = name, consumer, consumables_to_create
  end

  def produce!
    @consumables_to_create.times do |i|
      sleep((i % 2) / 1000)
      @consumer.add_consumable("#{@name} - #{i}")
    end
  end
end

consumer = Consumer.new
consumer_thread = Thread.new { consumer.consume! }

threads = []
threads << Thread.new { Producer.new("happy", consumer, 5).produce! }
threads << Thread.new { Producer.new("dopey", consumer, 8).produce! }
threads << Thread.new { Producer.new("skidmark", consumer, 10).produce! }

threads.each { |t| t.join }
# consumer_thread.join # deadlock
